// nodemon/lib/utils/log.js
var colour = require('./colour');
var bus = require('./bus');
var required = false;
var useColours = true;
var coding = {
  log: 'black',
  info: 'yellow',
  status: 'green',
  detail: 'yellow',
  fail: 'red',
  error: 'red',
};
function log(type, text) {
  var msg = '[nodemon] ' + (text || '');
  if (useColours) {
    msg = colour(coding[type], msg);
  }
  process.nextTick(() => {
    bus.emit('log', { type: type, message: text, colour: msg });
  });
  if (!required) {
    if (type === 'error') {
      console.error(msg);
    } else {
      console.log(msg || '');
    }
  }
}
var Logger = function (r) {
  if (!(this instanceof Logger)) {
    return new Logger(r);
  }
  this.required(r);
  return this;
};
Object.keys(coding).forEach(function (type) {
  Logger.prototype[type] = log.bind(null, type);
});
Logger.prototype.detail = function (msg) {
  if (this.debug) {
    log('detail', msg);
  }
};
Logger.prototype.required = function (val) {
  required = val;
};
Logger.prototype.debug = false;
Logger.prototype._log = function (type, msg) {
  if (required) {
    bus.emit('log', { type: type, message: msg || '', colour: msg || '' });
  } else if (type === 'error') {
    console.error(msg);
  } else {
    console.log(msg || '');
  }
};
Object.defineProperty(Logger.prototype, 'useColours', {
  set: function (val) {
    useColours = val;
  },
  get: function () {
    return useColours;
  },
});
module.exports = Logger;
